### ISSUE TITLE
Crash found with Fuzzing, preEncode()

### ISSUE THREAD

**Issue by nolanmar511** [google/pprof#245]
Please answer these questions before submitting your issue. Thanks!

### What version of pprof are you using?
commit: 9181e6bcbfdb5cd75bb563332b682b15b0b371b0 

### What operating system and processor architecture are you using?
Linux

### What did you do?
Modified the fuzzer to be :
```
func Fuzz(data []byte) int {
	p, err := profile.ParseData(data)
	if err == nil {
		if _, err := profile.Merge([]*profile.Profile{p,p}); err != nil {
			panic(fmt.Errorf("could not merge profile with self: %v", err))
		}
	}

	// Decompress if compressed.
	if len(data) >= 2 && data[0] == 0x1f && data[1] == 0x8b {
		gz, err := gzip.NewReader(bytes.NewBuffer(data))
		if err == nil {
			data, err = ioutil.ReadAll(gz)
		}
		if err != nil {
			return 0
		}
	}

	// if legacy format, just write out
	if _, err = profile.ParseUncompressed(data); err != nil {
		buf := bytes.NewBuffer(nil);
		p.Write(buf);
		return 0;
	}

	// if non-legacy format, write and compare to input
	buf := bytes.NewBuffer(nil);
	p.Write(buf);
	reEncoded := buf.Bytes();

	if diff, equal := messagediff.PrettyDiff(reEncoded, data); equal {
		panic(fmt.Errorf("Input doesn't match output: \n%s", diff));
	}

	return 0
}

```
Here is a zipped version of the profile which caused the crash. 
[c84dcaabd0c1937534e5488788e175e2bb066f3e.zip](https://github.com/google/pprof/files/1419740/c84dcaabd0c1937534e5488788e175e2bb066f3e.zip)

### What did you expect to see?
I expected no segmentation fault.

### What did you see instead?
There was a segmentation fault when writing out the parsed profile:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5110ba]

goroutine 1 [running]:
github.com/google/pprof/profile.(*Profile).preEncode(0xc4200aba40)
	/tmp/go-fuzz-build891219211/gopath/src/github.com/google/pprof/profile/encode.go:81 +0x118a
github.com/google/pprof/profile.serialize(0xc4200aba40, 0x5905d8, 0xc420096420, 0x40fe88)
	/tmp/go-fuzz-build891219211/gopath/src/github.com/google/pprof/profile/profile.go:309 +0x63
github.com/google/pprof/profile.(*Profile).Write(0xc4200aba40, 0x636360, 0xc4200cfc70, 0x0, 0x0)
	/tmp/go-fuzz-build891219211/gopath/src/github.com/google/pprof/profile/profile.go:319 +0x97
github.com/google/pprof/fuzz.Fuzz(0x7fe9cb5f8000, 0x10, 0x200000, 0xc420051f70)
	/tmp/go-fuzz-build891219211/gopath/src/github.com/google/pprof/fuzz/main.go:41 +0x1f3
go-fuzz-dep.Main(0x590648)
	/tmp/go-fuzz-build891219211/goroot/src/go-fuzz-dep/main.go:49 +0xad
main.main()
	/tmp/go-fuzz-build891219211/gopath/src/github.com/google/pprof/fuzz/go.fuzz.main/main.go:10 +0x2d
exit status 2

**Comment by aalexand** [comment]
I get a different stack with this crasher than what you wrote, see below. And it seems to be the cause similar to what I mentioned on #246. How do I reproduce what you are seeing?

```
=== RUN   TestCrashers/c84dcaabd0c1937534e5488788e175e2bb066f3e
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x517dc6]

goroutine 6 [running]:
testing.tRunner.func1(0xc4200c01e0)
	/usr/lib/google-golang/src/testing/testing.go:711 +0x2d2
panic(0x54d0c0, 0x634e50)
	/usr/lib/google-golang/src/runtime/panic.go:491 +0x283
github.com/google/pprof/profile.serialize(0x0, 0x5801e0, 0xc420084210, 0x410708)
	/usr/local/google/home/aalexand/go/src/github.com/google/pprof/profile/profile.go:308 +0x26
github.com/google/pprof/profile.(*Profile).Write(0x0, 0x6253a0, 0xc4200b5c70, 0x0, 0x0)
	/usr/local/google/home/aalexand/go/src/github.com/google/pprof/profile/profile.go:319 +0x7f
github.com/google/pprof/fuzz.Fuzz(0xc4200c4000, 0x10, 0x210, 0x10)
	/usr/local/google/home/aalexand/go/src/github.com/google/pprof/fuzz/main.go:42 +0x11a
github.com/google/pprof/fuzz.TestCrashers.func1(0xc4200c01e0)
	/usr/local/google/home/aalexand/go/src/github.com/google/pprof/fuzz/fuzz_test.go:26 +0x1d8
testing.tRunner(0xc4200c01e0, 0xc420058810)
	/usr/lib/google-golang/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/lib/google-golang/src/testing/testing.go:789 +0x2de
exit status 2
FAIL	github.com/google/pprof/fuzz	0.036s
```

**Comment by nolanmar511** [comment]
Sorry -- just included the stack returned when this profile failed in the fuzzer. 
And yes -- seems like crasher was produced due to same error in #246.

**Comment by aalexand** [comment]
Just to clarify: if the profile crashed in the fuzzer I would expect the same stack to reproduce with the same code and same data. Could it be that it was a different profile?

**Comment by nolanmar511** [comment]
Yes -- I think there was a crasher due to a function being nil, and it crashed in pre-encode. I may have deleted that profile, though, cause I cannot find it now. I'm re-running the fuzzer (and now returning when the profile is invalid, rather than trying to use it), but this would've been fixed by #248.
